src/rgw/rgw_rate_limit.cc

```diff
@@ -0,0 +1,373 @@
+}
+
+static api_counter_t *parse_limit(const YAML::Node& limit_data)
+{
+  api_counter_t *api_ctr = new api_counter_t();
+  try {
+    if ( !limit_data.size() ) {
+      delete api_ctr;
+      return NULL;
+    }
+
+    unsigned long long limit;
+    limit_data["limit"] >> limit;
+    api_ctr->limit.set(limit);
+    string period_str;
+    limit_data["period"] >> period_str;
+    if ( (api_ctr->period = parse_period_str(period_str)) == UNDEF ) {
+      delete api_ctr;
+      return NULL;
+    }
+
+  } catch (...) {
+    delete api_ctr;
+    return NULL;
+  }
+  return api_ctr;
+}
+
+static void swap_ctr(api_counter_t *old_ctr, api_counter_t *new_ctr)
+{
+  old_ctr->limit.set(new_ctr->limit.read());
+  old_ctr->period = new_ctr->period;
+}
+
+// Parse config file into tmp map & swap with current map
+static int load_limit_config_file()
+{
+  YAML::Node doc;
+  try {
+    ifstream config_file(RATE_LIMIT_CONFIG_FILE);
+    YAML::Parser parser(config_file);
+    parser.GetNextDocument(doc);
+  } catch (...) {
+    dout(0) << "Error parsing rate limit config" << dendl;
+    return -1;
+  }
+
+  dout(0) << "*** Rate limit config parsing ***" << dendl;
+
+  for ( unsigned int i=0; i < doc.size(); i++ ) {
+    string user;
+    const YAML::Node& node = doc[i];
+
+    try {
+      node["user"] >> user;
+    } catch (...) {
+      dout(0) << "Key 'user' not found, skipping stanza" << dendl;
+      continue;
+    }
+
+    api_counter_t *get_ctr;
+    try {
+      const YAML::Node& get = node["get"];
+      if ( (get_ctr = parse_limit(get)) == NULL ) {
+	throw;
+      } else {
+	api_ctr_it iter = rgw_get_ctrs->find(user);
+	if ( iter == rgw_get_ctrs->end() ) {
+	  (*rgw_get_ctrs)[user] = get_ctr;
+	} else {
+	  api_counter_t *ctr = iter->second;
+	  swap_ctr(ctr, get_ctr);
+	  delete get_ctr;
+	}
+      }
+    } catch (...) {
+      dout(0) << "Error parsing GET limit for user: " << user << dendl;
+    }
+
+    api_counter_t *put_ctr;
+    try {
+      const YAML::Node& put = node["put"];
+      if ( (put_ctr = parse_limit(put)) == NULL ) {
+	throw;
+      } else {
+	api_ctr_it iter = rgw_put_ctrs->find(user);
+	if ( iter == rgw_put_ctrs->end() ) {
+	  (*rgw_put_ctrs)[user] = put_ctr;
+	} else {
+	  api_counter_t *ctr = iter->second;
+	  swap_ctr(ctr, put_ctr);
+	  delete put_ctr;
+	}
+      }
+    } catch (...) {
+      dout(0) << "Error parsing PUT limit for user: " << user << dendl;
+    }
+
+    if ( get_ctr ) {
+      dout(0) << "User: " << user \
+	      << "\tGET: " << get_ctr->limit.read() \
+	      << "/" << get_period_str(get_ctr->period) << dendl;
+    }
+    if ( put_ctr ) {
+      dout(0) << "User: " << user \
+	      << "\tPUT: " << put_ctr->limit.read() \
+	      << "/" << get_period_str(put_ctr->period) << dendl;
+    }
+  }
+
+  dout(0) << "*** Final GET rate limit config ***" << dendl;
+  dump_limit_config(rgw_get_ctrs);
+  dout(0) << "*** Final PUT rate limit config ***" << dendl;
+  dump_limit_config(rgw_put_ctrs);
+
```

Comment:
The unordered map may have newer entries (more likely) or existing entries deleted (we need not delete entries so this won't be of concern).  And we don't know how that would've been implemented.